packages/compose/src/hooks/use-observe-element-size/index.ts

```diff
@@ -0,0 +1,116 @@
+/**
+ * External dependencies
+ */
+import type { RefObject } from 'react';
+/**
+ * WordPress dependencies
+ */
+import { useRef, useEffect, useLayoutEffect } from '@wordpress/element';
+/**
+ * Internal dependencies
+ */
+import useEvent from '../use-event';
+
+/**
+ * Tracks a given element's size and calls `onUpdate` for all of its discrete
+ * values using a `ResizeObserver`. The element can change dynamically. Alternatively,
+ * a ref containing the target element can be passed. However, any updates to the ref
+ * will be ignored.
+ *
+ * @param targetElement         The target element to observe. It can be changed dynamically. Alternatively, a ref containing the target element can be passed. However, any updates to the ref will be ignored.
+ * @param onUpdate              Callback that will be called when the element is resized. It is passed the list of [`ResizeObserverEntry`](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry) objects passed to the `ResizeObserver.observe` callback internally, and the element being tracked at the time of this update.
+ * @param resizeObserverOptions Options to pass to `ResizeObserver.observe` when called internally. Updating this option will not cause the observer to be re-created, and it will only take effect if a new element is observed.
+ *
+ * @example
+ *
+ * ```tsx
+ * const [ targetElement, setTargetElement ] = useState< HTMLElement | null >();
+ * useObserveElementSize(
+ * 	targetElement,
+ * 	( resizeObserverEntries, element ) => {
+ * 		console.log( 'Resize observer entries:', resizeObserverEntries );
+ * 		console.log( 'Element that was resized:', element );
+ * 	},
+ * 	{ box: 'border-box' }
+ * );
+ * <div ref={ setTargetElement } />;
+ * ```
+ */
+export default function useObserveElementSize(
+	/**
+	 * The target element to observe. It can be changed dynamically.
+	 *
+	 * Alternatively, a ref containing the target element can be passed. However, any updates
+	 * to the ref will be ignored.
+	 */
+	targetElement: HTMLElement | undefined | null | RefObject< HTMLElement >,
```

Comment:
This could be simply a `RefObject< HTMLElement >`? The use case where you have a direct `targetElement` value can be reduced to passing `{ current: targetElement }`.

Reply:
No. that doesn't work because element needs to be tracked when not passed in a ref and it needs to be part of React's scheduling. If a ref is passed, then `targetElement` (in ref form) needs to be stable when passed to the effect dep array.

Reply:
Actually, I personally don't like hooks that receive refs as argument. I think it's an anti-pattern that we should move away from. The reason is simple, quickly we'll found ourselves with bugs because the ref changes over time but the hook doesn't trigger again (you can't watch a ref).

So IMO, we should refactor the code to return the "ref" instead of retrieving it.

Reply:
Returning the ref is no different. You've got the exact same issue regardless of whether you create the ref yourself get it from the hook. The only way to "watch" an element is to have it be a state, and that is supported by this hook.

However, it is true that using refs for elements is quite common, and while if I had my way I'd always enforce elements in the state, @jsnajdr has a point about it being a common React idiom. So this hook supports them, while clearly stating that updates won't be tracked (this should also not be news to the average React developer). And it also offers a stateful counterpart, as it's necessary for my use case(s). It doesn't get more straightforward than this. Returning the ref would make the stateful version incompatible, which doesn't seem worth it for the sake of saving the user one line of code (`useRef` call).

This is the absolutely simplest API and implementation for a React `ResizeObserver` abstraction, and any further API abstractions or omissions only result in blocking a subset of use-cases and forcing forks down the road, as I argued previously in this PR.

Can we please compromise on this API, which supports all use-cases and is thoroughly documented? Any other APIs can be achieved with this one thanks to composition, so it won't ever block those. This is not true the other way around, because [composability is unidirectional](https://dio.la/article/the-everything-bagel-of-components#the-in-between-composability-is-unidirectional).

Reply:
> Returning the ref is no different. You've got the exact same issue regardless of whether you create the ref yourself get it from the hook. The only way to "watch" an element is to have it be a state, and that is supported by this hook.

This is not true.

if you pass `const ref = useRef()` 

you won't be able to watch the ref but if you do

```
const myRef = useObserveElementSize( callback );

return someCondition ? <div ref={myRef} /> : <span ref={myRef} />;
```

The hook can observe the ref properly and adapt to the Dom element change. This is why we have a `useRefEffect` hook.

Reply:
Well, fair enough. Looking at `useRefEffect`, you're returning a callback ref and use `useCallback` to track dependencies statefully. To be honest, I prefer to use a state for stateful stuff, rather than this which feels like extra steps and kinda magic from the outside. That said, I see how this could be a valid way to observe changes in a ref, so fair enough.

With all of this in mind, my use case is still necessary (simply passing an element obtained in different ways, including callback refs and `document.querySelector`), and in fact a hook that implements the API you suggest could be created from mine with ease (but not viceversa).

So can we please compromise on this to get it merged? Then, if you feel like you need an intermediate hook that returns a ref and uses this trick, you can always create one from mine. Though given that the higher level `useResizeObserver` utility is already consistently used everywhere, I doubt it'll be necessary.

Reply:
> So can we please compromise on this to get it merged? Then, if you feel like you need an intermediate hook that returns a ref and uses this trick, you can always create one from mine. Though given that the higher level useResizeObserver utility is already consistently used everywhere, I doubt it'll be necessary.

You're putting me in a bad place here. I don't want to be the harsh guy that blocks everything but at the same time, we started with similar to this kind of APIs in the past (for a lot of the existing hooks that use useRefEffect) and we decided to slowly move away from them and the reason for this is still valid. So shipping this would be reverting that, so let's see first if we really need to revert that decision or we can keep following our best practice.

> With all of this in mind, my use case is still necessary (simply passing an element obtained in different ways, including callback refs and document.querySelector), and in fact a hook that implements the API you suggest could be created from mine with ease (but not viceversa).

Can you show me where you want to use this hook in Gutenberg to see if we can solve it differently?

Reply:
A quick update. After some discussion with @jsnajdr (https://github.com/WordPress/gutenberg/pull/64943#discussion_r1745658233), he's given floating-ui as an example, and I really like what they do: they return a callback ref but also allow passing an element directly. This should cover all of our preferences while remaining reactive to element changes. I will also use `useRefEffect` as inspiration to save a re-render.

Win-win-win? (please? :D)

Will push latest version in a bit.

Reply:
> we started with similar to this kind of APIs in the past (for a lot of the existing hooks that use useRefEffect) and we decided to slowly move away from them

@youknowriad Can you share an example of such an API, what problems it caused and how we moved away from it? Correctly tracking elements in refs is a hairy problem that I still don't fully understand, and would like to see something that would clarify it for me.

I've been looking closely at `useFloating` now and it can be used like:
```js
const { refs } = useFloating();
return <div ref={ refs.setFloating } />;
```
But internally it stores the `floatingEl` in state and is going to do a `setState`. Which is not needed if I'm sure that the rendered `<div>` is stable and constant.

But without `setState`, I can't reliably detect when the element changes or is rendered conditionally:
```js
return condition ? <div ref={ setFloating } /> : <span ref={ setFloating } />;
```

The current implementation of `useObserveElementSize` assumes that the ref is set on the initial mount on the hook and then remains stable forever. That's true for most usages, and I'm personally fine with that, but it might be confusing.

Reply:
The TLDR is something like this

```js
function useAttachSomeEventHandler( ref ) {
   useEffect(() => {
      const callback =  () => {
          // Do something when the node is clicked
      };
      ref.current.addEventListener( 'click', callback );
      return () => ref.current.removeEventListener( 'click', callback );
   }, [] );
}

function MyComponent({ someCondition }) { 
  const ref = useRef() 
  useAttachSomeEventHandler( ref );
  return someCondition ? <div ref={ref} /> : <span ref={ref} />;
}
```

In the example above, if "someCondition" changes during runtime, the event listener is not attached anymore to the right DOM element.

While I have to look back to some of @ellatrix's PR mostly, I know we have similar issues like that for hooks that are applied to `useBlockProps` in the past.

Reply:
Oh yes, that's exactly the behavior I labeled as acceptable in my previous comment 🙂 

Is it better if I return the result of `useRefEffect` like here?
```js
function useObserveElementSize( onResize ) {
  const onResizeCb = useEvent( onResize );

  return useRefEffect( ( element ) => {
    const observer = new ResizeObserver( onResizeCb );
    observer.observe( element );
    return () => observer.unobserve( element );
  } );
}
```
Now the remaining task is to add the option to pass `element` directly. Maybe the hook could have just one param with options, some mandatory, some not:
```js
useObserveElementSize( { onResize, element: myElement, options: optionsForResizeObserver } );
```

> Can we rename it `useResizeObserver` and kind of detect whether we want the new or old version

The original version has no arguments at all, while the new version has always at least one, so it's easy to distinguish between them.

Reply:
> Is it better if I return the result of useRefEffect like here?

Well if you return a useRefEffect, there's no bug anymore, when the DOM element changes, the event will be removed from the previous element and attached to the new one.

> That's true for most usages, and I'm personally fine with that, but it might be confusing.

Personally, I'm not fine with that. I'll give you a concrete example:

 - Let's say we attach a behavior to a block in `useBlockProps`. For example I pass a `ref` to it in order to check the block size/position. 
 - Some blocks have conditional rendering like rendering a div placeholder when empty and a "figure" when we fill an image for instance.
 - As soon as the switch happens, the bug happens, the ref is not computing the size properly.

Reply:
Thanks for the discussion folks.

Check out the latest implementation. I believe I covered everything:

- Supports legacy API.
- Has proper overloads: if you use the old API you still get all of the docs, with a note recommending the new one though.
- Docs output the new API.
- The returned value is a callback ref which means there's no need for state. I will always defend the idea that re-renders are basically free but in this case this is a clean enough solution anyway.
- Supports alternative element input just like the floating UI API, which neatly enables my use case without fragile hacks.
- No chance for de-sync. You'd have to try hard to make an observer instance stale. `useRefEffect`-inspired!

I agree with @youknowriad in all accounts, which is the reason why I was initially only supporting a direct (stateful) element input. That said, the callback ref return + the alternative seems like a good balance that should make us all happy and optimize for the most common path while enabling more advanced uses, and all without a chance for missed updates.

Looking forward to your reviews!